File: Two Pointers/Medium/LET_no5_Longest Palindromic Substring(middle_enumeration).py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        left = right = 0
        for k in range(len(s)):
            i, j = self.helper(s, k, k)
            if j - i > right - left:
                left, right = i, j
            
            if k > 0:
                i, j = self.helper(s, k - 1, k)
                if j - i > right - left:
                    left, right = i, j
        
        return s[left:right+1]
    
    def helper(self, s, left, right):
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return left + 1, right - 1

File: Two Pointers/Medium/LET_no5_Longest Palindromic Substring(middle_enumeration).cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int res_left = 0, res_right = 0;
        for (int i = 0; i < s.size(); i++) {
            pair<int, int> interval = GetSubstringInterval(s, i, i);
            if (interval.second - interval.first > res_right - res_left) {
                res_left = interval.first;
                res_right = interval.second;
            }

            if (i > 0) {
                pair<int, int> interval = GetSubstringInterval(s, i - 1, i);
                if (interval.second - interval.first > res_right - res_left) {
                    res_left = interval.first;
                    res_right = interval.second;
                }
            }
        }

        return s.substr(res_left, res_right - res_left + 1);
    }

    pair<int, int> GetSubstringInterval(string& s, int left, int right) {
        int n = s.size();
        while (left >= 0 && right < n && s[left] == s[right]) {
            left--;
            right++;
        }
        
        return {left + 1, right - 1};
    }
};